src/ChoppingKinky.cpp

```diff
@@ -0,0 +1,362 @@
+
+		// if the CHOPP gate is wired in, do chop logic
+		if (inputs[IN_GATE_INPUT].isConnected()) {
+			// TODO: check rescale?
+			trigger.process(rescale(inputs[IN_GATE_INPUT].getVoltage(), 0.1f, 2.f, 0.f, 1.f));
+			outputAToChopp = trigger.isHigh();
+		}
+		// else zero-crossing detector on input A switches between A and B
+		else {
+			if (previousA > 0 && inA < 0) {
+				outputAToChopp = false;
+			}
+			else if (previousA < 0 && inA > 0) {
+				outputAToChopp = true;
+			}
+		}
+		previousA = inA;
+
+		const bool choppIsRequired = outputs[OUT_CHOPP_OUTPUT].isConnected();
+		const bool aIsRequired = outputs[OUT_A_OUTPUT].isConnected() || choppIsRequired;
+		const bool bIsRequired = outputs[OUT_B_OUTPUT].isConnected() || choppIsRequired;
+
+		if (aIsRequired) {
+			oversampler[CHANNEL_A].upsample(inA * gainA);
+		}
+		if (bIsRequired) {
+			oversampler[CHANNEL_B].upsample(inB * gainB);
+		}
+		if (choppIsRequired) {
+			oversampler[CHANNEL_CHOPP].upsample(outputAToChopp ? 1.f : 0.f);
+		}
+
+		float* osBufferA = oversampler[CHANNEL_A].getOSBuffer();
+		float* osBufferB = oversampler[CHANNEL_B].getOSBuffer();
+		float* osBufferChopp = oversampler[CHANNEL_CHOPP].getOSBuffer();
+
+		for (int i = 0; i < oversampler[0].getOversamplingRatio(); i++) {
+			if (aIsRequired) {
+				//osBufferA[i] = wavefolderAResponse(osBufferA[i]);
+				osBufferA[i] = wavefolderAResponseCached(osBufferA[i]);
+			}
+			if (bIsRequired) {
+				//osBufferB[i] = wavefolderBResponse(osBufferB[i]);
+				osBufferB[i] = wavefolderBResponseCached(osBufferB[i]);
+			}
+			if (choppIsRequired) {
+				osBufferChopp[i] = osBufferChopp[i] * osBufferA[i] + (1.f - osBufferChopp[i]) * osBufferB[i];
+			}
+		}
+
+		float outA = aIsRequired ? oversampler[CHANNEL_A].downsample() : 0.f;
+		float outB = bIsRequired ? oversampler[CHANNEL_B].downsample() : 0.f;
+		float outChopp = choppIsRequired ? oversampler[CHANNEL_CHOPP].downsample() : 0.f;
+
+		if (blockDC) {
+			outChopp = blockDCFilter.process(outChopp);
+		}
+
+		outputs[OUT_A_OUTPUT].setVoltage(outA);
+		outputs[OUT_B_OUTPUT].setVoltage(outB);
+		outputs[OUT_CHOPP_OUTPUT].setVoltage(outChopp);
+
+		if (inputs[IN_GATE_INPUT].isConnected()) {
+			lights[LED_A_LIGHT].setSmoothBrightness((float) outputAToChopp, args.sampleTime);
+			lights[LED_B_LIGHT].setSmoothBrightness((float)(!outputAToChopp), args.sampleTime);
+		}
+		else {
+			lights[LED_A_LIGHT].setBrightness(0.f);
+			lights[LED_B_LIGHT].setBrightness(0.f);
+		}
+	}
+
+	float wavefolderAResponseCached(float x) {
+		if (x >= 0) {
+			float j = rescale(clamp(x, 0.f, 10.f), 0.f, 10.f, 0, WAVESHAPE_CACHE_SIZE - 1);
+			return interpolateLinear(waveshapeA, j);
+		}
+		else {
+			return -wavefolderAResponseCached(-x);
+		}
+	}
+
+	float wavefolderBResponseCached(float x) {
+		if (x >= 0) {
+			float j = rescale(clamp(x, 0.f, 10.f), 0.f, 10.f, 0, WAVESHAPE_CACHE_SIZE - 1);
+			return interpolateLinear(waveshapeBPositive, j);

```

Comment:
@squinkylabs - there is linear interpolation here. I've put audio through it and looked at the spectrum and everything looks to be OK?

top is no anti-aliasing and below is 16x: 
![image](https://user-images.githubusercontent.com/915048/124825966-309d2880-df6c-11eb-8807-84c0fab12335.png)

Reply:
I guess I'm not sure what truncating to 8 bits looks like on the spectrum analyzer, but I guess it would be low level junk? The spectra do look good - I guess the input here is a 1 k sine? I think I would expect so see weird non-harmonic frequencies below -48 db? Is there a setting where the signal is run through the table, but the table is (almost) a straight line? Then I think the effect of the table would be to truncate your 32 bit audio down to 8 bits. Looks like the oversampling is working well - nice.

Reply:
Not quite what you had in mind, but if you "un-interpolate" it looks like this:

![image](https://user-images.githubusercontent.com/915048/124829997-38ab9700-df71-11eb-9653-1fdba8172550.png)


```float wavefolderBResponseCached(float x) {
		if (x >= 0) {
			float j = std::round(rescale(clamp(x, 0.f, 10.f), 0.f, 10.f, 0, WAVESHAPE_CACHE_SIZE - 1));
			return interpolateLinear(waveshapeBPositive, j);
		}
		else {
			float j = std::round(rescale(clamp(-x, 0.f, 10.f), 0.f, 10.f, 0, WAVESHAPE_CACHE_SIZE - 1));
			return interpolateLinear(waveshapeBNegative, j);
		}
	}
```

Reply:
what am I seeing here? the two on the left look good, and the two on the right look bad.. which ones have your interpolation? and was that interpolation always in there and I missed it?

Reply:
Sorry - yes interpolation was always there. I suspected it would be pretty obvious visually and in the specturm, but I wanted to check. Left is PR version (top no AA, bottom 16x), right is what would happen without interpolation (I just temporarily patched it to see).

Reply:
oh! ok, sorry I missed that. yeah, looks like a) I was wrong about your code, b) the interpolation looks good, c) hey, at least I was in the ballpark for guessing how much junk would be there! super nice work here, and apologies again for missing that.

Reply:
All good - your DSP guides have been super useful during this process, so thanks for that!